File: app/src/main/java/com/michel/deutschmacht/data/Lessons.java
package com.michel.deutschmacht.data;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/** Loads lessons.json from assets. Duplicated booklet lines are intentionally kept. */
public final class Lessons {

    public static class Entry {
        public final String de;
        public final String fa;
        public final int lesson;   // 1-based lesson number
        public final int index;    // position inside the lesson

        Entry(String de, String fa, int lesson, int index) {
            this.de = de;
            this.fa = fa;
            this.lesson = lesson;
            this.index = index;
        }
    }

    public static class Lesson {
        public final int num;
        public final List<Entry> entries = new ArrayList<>();

        Lesson(int num) { this.num = num; }
    }

    private static List<Lesson> cache;

    private Lessons() {}

    public static synchronized List<Lesson> load(Context ctx) {
        if (cache != null) return cache;
        List<Lesson> out = new ArrayList<>();
        try {
            StringBuilder sb = new StringBuilder();
            try (BufferedReader r = new BufferedReader(
                    new InputStreamReader(ctx.getAssets().open("lessons.json"), StandardCharsets.UTF_8))) {
                String line;
                while ((line = r.readLine()) != null) sb.append(line);
            }
            JSONArray root = new JSONArray(sb.toString());
            for (int i = 0; i < root.length(); i++) {
                JSONObject lo = root.getJSONObject(i);
                Lesson l = new Lesson(lo.getInt("num"));
                JSONArray arr = lo.getJSONArray("entries");
                for (int j = 0; j < arr.length(); j++) {
                    JSONObject eo = arr.getJSONObject(j);
                    l.entries.add(new Entry(eo.getString("de"), eo.getString("fa"), l.num, j));
                }
                out.add(l);
            }
        } catch (Exception e) {
            // Never crash the app over data: leave an empty structure.
        }
        cache = out;
        return out;
    }

    public static Lesson lesson(Context ctx, int num) {
        for (Lesson l : load(ctx)) if (l.num == num) return l;
        return new Lesson(num);
    }

    public static int totalEntries(Context ctx) {
        int n = 0;
        for (Lesson l : load(ctx)) n += l.entries.size();
        return n;
    }

    public static List<Entry> flat(Context ctx) {
        List<Entry> all = new ArrayList<>();
        for (Lesson l : load(ctx)) all.addAll(l.entries);
        return all;
    }
}

File: app/src/main/java/com/michel/deutschmacht/data/Lessons.kt
package com.michel.deutschmacht.data

import android.content.Context
import org.json.JSONArray
import java.io.BufferedReader
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

/**
 * Loads [lessons.json] from assets. Duplicated booklet lines are intentionally kept:
 * the repetition itself is the Michel Thomas method.
 */
object Lessons {

    data class Entry(val de: String, val fa: String, val lesson: Int, val index: Int)

    data class Lesson(val num: Int, val entries: MutableList<Entry> = mutableListOf())

    @Volatile
    private var cache: List<Lesson>? = null

    private val lock = Any()

    fun load(context: Context): List<Lesson> = synchronized(lock) {
        cache?.let { return it }
        val parsed = try {
            parse(context)
        } catch (e: Exception) {
            emptyList()
        }
        cache = parsed
        parsed
    }

    private fun parse(context: Context): List<Lesson> {
        val sb = StringBuilder()
        BufferedReader(InputStreamReader(context.assets.open("lessons.json"), StandardCharsets.UTF_8)).use { r ->
            var line = r.readLine()
            while (line != null) {
                sb.append(line)
                line = r.readLine()
            }
        }
        val out = mutableListOf<Lesson>()
        val root = JSONArray(sb.toString())
        for (i in 0 until root.length()) {
            val lo = root.getJSONObject(i)
            val lesson = Lesson(lo.getInt("num"))
            val arr = lo.getJSONArray("entries")
            for (j in 0 until arr.length()) {
                val eo = arr.getJSONObject(j)
                lesson.entries.add(Entry(eo.getString("de"), eo.getString("fa"), lesson.num, j))
            }
            out.add(lesson)
        }
        return out
    }

    fun lesson(context: Context, num: Int): Lesson =
        load(context).firstOrNull { it.num == num } ?: Lesson(num)

    fun totalEntries(context: Context): Int = load(context).sumOf { it.entries.size }

    fun flat(context: Context): List<Entry> = load(context).flatMap { it.entries }
}
